pkg/component/controller/etcd_member_reconciler.go

```diff
@@ -56,21 +60,57 @@ type EtcdMemberReconciler struct {
 	etcdConfig       *v1beta1.EtcdConfig
 	etcdMemberClient etcdmemberclient.EtcdMemberInterface
 	leaderElector    leaderelector.Interface
+	mux              sync.Mutex
+	started          atomic.Bool
 }
 
 func (e *EtcdMemberReconciler) Init(_ context.Context) error {
 	return nil
 }
 
+// resync does a full resync of the etcd members when the leader changes
+// This is needed to ensure all the member objects are in sync with the actual etcd cluster
+// We might get stale state if we remove the current leader as the leader will essentially
+// remove itself from the etcd cluster and after that tries to update the member object.
+func (e *EtcdMemberReconciler) resync(ctx context.Context) error {
+	e.mux.Lock()
+	defer e.mux.Unlock()
+
+	if !e.started.Load() {
+		logrus.WithField("component", "EtcdMemberReconciler").Debug("Not started yet!!?!?")
+		return nil
+	}
+
+	// Loop through all the members and run reconcile on them
+	// Use high timeout as etcd/api could be a bit slow when the leader changes
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+	members, err := e.etcdMemberClient.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, member := range members.Items {
+		e.reconcileMember(ctx, &member)
+	}
+	return nil
+}
+
 func (e *EtcdMemberReconciler) Start(ctx context.Context) error {
-	log := logrus.WithField("component", "etcdMemberReconciler")
+	log := logrus.WithField("component", "EtcdMemberReconciler")
 
 	etcdMemberClient, err := e.clientFactory.GetEtcdMemberClient()
 	if err != nil {
 		return err
 	}
 	e.etcdMemberClient = etcdMemberClient
 
+	e.leaderElector.AddAcquiredLeaseCallback(func() {
+		log.Info("leader lease acquired, starting resync")
+		if err := e.resync(ctx); err != nil {
```

Comment:
It's not a good idea to fully block the callback here. The current leader elector implementation won't call each callback in its own goroutine, unfortunately. So the resync here has the possibility of blocking the whole leader election processing for the whole k0s controller, especially as the resync func takes a mutex lock.

(We definitely need a better leader elector component with a nicer API that's more resilient.)

Reply:
Good catch, this is a detail I always seem to forget. I'll make it spin up a routine